File: commons-core/src/main/java/xyz/wagyourtail/commons/position/Pos2.java
package xyz.wagyourtail.commons.position;

import lombok.AllArgsConstructor;

import java.util.Objects;

@AllArgsConstructor
public class Pos2 {
    public static final Pos2 ZERO = new Pos2(0, 0);

    public final int x;
    public final int y;

    public static Pos2 max(Pos2 p1, Pos2 p2) {
        return new Pos2(Math.max(p1.x, p2.x), Math.max(p1.y, p2.y));
    }

    public static Pos2 min(Pos2 p1, Pos2 p2) {
        return new Pos2(Math.min(p1.x, p2.x), Math.min(p1.y, p2.y));
    }

    public static Pos2 abs(Pos2 p) {
        return new Pos2(Math.abs(p.x), Math.abs(p.y));
    }

    public Pos2 up() {
        return new Pos2(x, y - 1);
    }

    public Pos2 down() {
        return new Pos2(x, y + 1);
    }

    public Pos2 left() {
        return new Pos2(x - 1, y);
    }

    public Pos2 right() {
        return new Pos2(x + 1, y);
    }

    public Pos2 inverse() {
        return new Pos2(-x, -y);
    }

    public Pos2 plus(Pos2 other) {
        return new Pos2(x + other.x, y + other.y);
    }

    public Pos2 plus(int x, int y) {
        return new Pos2(x + this.x, y + this.y);
    }

    public Pos2 plus(int scalar) {
        return new Pos2(x + scalar, y + scalar);
    }

    public Pos2 minus(Pos2 other) {
        return new Pos2(x - other.x, y - other.y);
    }

    public Pos2 minus(int x, int y) {
        return new Pos2(x - this.x, y - this.y);
    }

    public Pos2 minus(int scalar) {
        return new Pos2(x - scalar, y - scalar);
    }

    public Pos2 times(Pos2 other) {
        return new Pos2(x * other.x, y * other.y);
    }

    public Pos2 times(int x, int y) {
        return new Pos2(x * this.x, y * this.y);
    }

    public Pos2 times(int scalar) {
        return new Pos2(x * scalar, y * scalar);
    }

    public Pos2 div(Pos2 other) {
        return new Pos2(x / other.x, y / other.y);
    }

    public Pos2 div(int x, int y) {
        return new Pos2(x / this.x, y / this.y);
    }

    public Pos2 div(int scalar) {
        return new Pos2(x / scalar, y / scalar);
    }

    public Pos2 rem(Pos2 other) {
        return new Pos2(x % other.x, y % other.y);
    }

    public Pos2 rem(int x, int y) {
        return new Pos2(this.x % x, this.y % y);
    }

    public Pos2 rem(int scalar) {
        return new Pos2(x % scalar, y % scalar);
    }

    public Pos2 mod(Pos2 other) {
        int x = this.x % other.x;
        int y = this.y % other.y;
        if (x < 0) x += other.x;
        if (y < 0) y += other.y;
        return new Pos2(x, y);
    }

    public Pos2 mod(int x, int y) {
        int x2 = this.x % x;
        int y2 = this.y % y;
        if (x2 < 0) x2 += x;
        if (y2 < 0) y2 += y;
        return new Pos2(x2, y2);
    }

    public Pos2 mod(int scalar) {
        int x = this.x % scalar;
        int y = this.y % scalar;
        if (x < 0) x += scalar;
        if (y < 0) y += scalar;
        return new Pos2(x, y);
    }

    public Pos2D toDouble() {
        return new Pos2D(x, y);
    }

    public double distanceTo(Pos2 other) {
        return Math.sqrt(distanceToSquared(other));
    }

    public int distanceToSquared(Pos2 other) {
        return (x - other.x) * (x - other.x) + (y - other.y) * (y - other.y);
    }

    public int manhattanDistanceTo(Pos2 other) {
        return Math.abs(x - other.x) + Math.abs(y - other.y);
    }

    public double magnitude() {
        return Math.sqrt(x * x + y * y);
    }

    public int magnitudeSquared() {
        return x * x + y * y;
    }

    public int manhattanMagnitude() {
        return Math.abs(x) + Math.abs(y);
    }

    public double dot(Pos2 other) {
        return x * other.x + y * other.y;
    }

    @Override
    public String toString() {
        return "(" + x + ", " + y + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Pos2)) return false;
        Pos2 pos2 = (Pos2) o;
        return x == pos2.x && y == pos2.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

}

File: commons-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/position/Pos2.kt
package xyz.wagyourtail.commonskt.position

import kotlin.math.sqrt

open class Pos2(val x: Int, val y: Int) {
    companion object {
        val ZERO = Pos2(0, 0)
    }

    open operator fun unaryMinus(): Pos2 {
        return Pos2(-x, -y)
    }

    open val up
        get() = Pos2(this.x, this.y - 1)

    open val down
        get() = Pos2(this.x, this.y + 1)

    open val left
        get() = Pos2(this.x - 1, this.y)

    open val right
        get() = Pos2(this.x + 1, this.y)

    open operator fun rangeTo(pos: Pos2): Sequence<Pos2> {
        return sequence {
            for (y in y..pos.y) {
                for (x in x..pos.x) {
                    yield(Pos2(x, y))
                }
            }
        }
    }

    open operator fun rangeUntil(pos: Pos2): Sequence<Pos2> {
        return sequence {
            for (y in y ..< pos.y) {
                for (x in x + 1 ..< x) {
                    yield(Pos2(x, y))
                }
            }
        }
    }

    operator fun plus(other: Pos2): Pos2 {
        return Pos2(x + other.x, y + other.y)
    }

    fun plus(x: Int, y: Int): Pos2 {
        return Pos2(x + this.x, y + this.y)
    }

    open operator fun plus(scalar: Int): Pos2 {
        return Pos2(x + scalar, y + scalar)
    }

    operator fun minus(other: Pos2): Pos2 {
        return Pos2(x - other.x, y - other.y)
    }

    fun minus(x: Int, y: Int): Pos2 {
        return Pos2(x - this.x, y - this.y)
    }

    open operator fun minus(scalar: Int): Pos2 {
        return Pos2(x - scalar, y - scalar)
    }

    operator fun times(other: Pos2): Pos2 {
        return Pos2(x * other.x, y * other.y)
    }

    fun times(x: Int, y: Int): Pos2 {
        return Pos2(x * this.x, y * this.y)
    }

    open operator fun times(scalar: Int): Pos2 {
        return Pos2(x * scalar, y * scalar)
    }

    operator fun div(other: Pos2): Pos2 {
        return Pos2(x / other.x, y / other.y)
    }

    fun div(x: Int, y: Int): Pos2 {
        return Pos2(x / this.x, y / this.y)
    }

    open operator fun div(scalar: Int): Pos2 {
        return Pos2(x / scalar, y / scalar)
    }

    operator fun rem(other: Pos2): Pos2 {
        return Pos2(x % other.x, y % other.y)
    }

    fun rem(x: Int, y: Int): Pos2 {
        return Pos2(this.x % x, this.y % y)
    }

    open operator fun rem(scalar: Int): Pos2 {
        return Pos2(x % scalar, y % scalar)
    }

    infix fun mod(other: Pos2): Pos2 {
        var x = this.x % other.x
        var y = this.y % other.y
        if (x < 0) x += other.x
        if (y < 0) y += other.y
        return Pos2(x, y)
    }

    fun mod(x: Int, y: Int): Pos2 {
        var x2 = this.x % x
        var y2 = this.y % y
        if (x2 < 0) x2 += x
        if (y2 < 0) y2 += y
        return Pos2(x2, y2)
    }

    open infix fun mod(scalar: Int): Pos2 {
        var x = this.x % scalar
        var y = this.y % scalar
        if (x < 0) x += scalar
        if (y < 0) y += scalar
        return Pos2(x, y)
    }

    open fun toDouble(): Pos2D {
        return Pos2D(x.toDouble(), y.toDouble())
    }

    open val magnitude: Double
        get() {
            return sqrt(magnitudeSquared.toDouble())
        }

    open val magnitudeSquared: Int
        get() {
            return x * x + y * y
        }

    fun dot(other: Pos2): Int {
        return x * other.x + y * other.y
    }

    override fun toString(): String {
        return "($x, $y)"
    }

    override fun equals(other: Any?): Boolean {
        return other is Pos2 && other.x == this.x && other.y == this.y
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + y.hashCode()
        return result
    }
}
